### PR TITLE
Add NIXL storage volume transport

### PR DESCRIPTION
**Pull request by tushar00jain**
- Add a NIXL based TransportBuffer. This ended up mostly a copy-paste of
  existing transports e.g. Torchcomms.
- NIXL is the first preffered transport but it remains disabled by
  default using existing env var convention

fs-branch: rdma/1

---

<sub>Stack created with <a href="https://github.com/github/gh-stack">GitHub Stacks CLI</a> • <a href="https://gh.io/stacks-feedback">Give Feedback 💬</a></sub>

### CODE REVIEW

File: torchstore/transport/nixl.py [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+from dataclasses import dataclass
+from functools import cache
+from typing import Any, TYPE_CHECKING
+
+import torch
+
+from torchstore.transport.buffers import TransportBuffer, TransportCache
+from torchstore.transport.types import Request
+
+if TYPE_CHECKING:
```
Comment (ivy-zhou) [inline torchstore/transport/nixl.py:27]: nit: I understand this is a problem in the other transport files as well but I feel like there should be a way to avoid this?

File: torchstore/transport/nixl.py [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+            from nixl_cu12 import nixl_agent_config as _nixl_agent_config
+        except ImportError:
+            _nixl_agent = None
+            _nixl_agent_config = None
+
+
+logger = logging.getLogger(__name__)
+
+
+def _backend() -> str:
```
Comment (ivy-zhou) [inline torchstore/transport/nixl.py:50]: why not top level defaulted variable? why a function?

File: torchstore/transport/nixl.py [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+            )
+            assert request_context.remote_descriptors is not None
+            transfers.append(
+                (request, tensor, request_context.remote_descriptors)
+            )
+            results.append(tensor)
+        await self._transfer(ctx, "READ", transfers)
+        return results
+
+    async def handle_get_request(
```
Comment (ivy-zhou) [inline torchstore/transport/nixl.py:368]: nit: would prefer that these are annotated with `@override`

File: torchstore/transport/nixl.py [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+    is_object: bool = False
+    objects: Any = None
+
+    def __getstate__(self) -> dict[str, Any]:
+        state = self.__dict__.copy()
+        state["tensor"] = None
+        return state
+
+
+class NixlTransportBuffer(TransportBuffer):
```
Comment (ivy-zhou) [inline torchstore/transport/nixl.py:165]: Can we get a test for this?

File: torchstore/transport/nixl.py [hunk 5]
```diff
@@ -0,0 +1,415 @@
...
+    remote_descriptors: bytes | None = None
+    tensor: torch.Tensor | None = None
+    shape: torch.Size | None = None
+    dtype: torch.dtype | None = None
+    is_object: bool = False
+    objects: Any = None
+
+    def __getstate__(self) -> dict[str, Any]:
+        state = self.__dict__.copy()
+        state["tensor"] = None
```
Comment (ivy-zhou) [inline torchstore/transport/nixl.py:161]: Worth a comment here that this is deleted because we don't want to send the tensor through RPC?

File: torchstore/transport/nixl.py [hunk 6]
```diff
@@ -0,0 +1,415 @@
...
+                remote_descs,
+                remote_agent,
+                backends=[cache.backend],
+            )
+            status = cache.agent.transfer(handle)
+
+            deadline = time.monotonic() + float(
+                os.environ.get("TORCHSTORE_NIXL_TIMEOUT_S", "60")
+            )
+            while status == "PROC" and time.monotonic() < deadline:
```
Comment (ivy-zhou) [inline torchstore/transport/nixl.py:312]: Can you define "PROC" and the other NIXL constants as constants in this file please?

File: torchstore/transport/nixl.py [hunk 7]
```diff
@@ -0,0 +1,415 @@
...
+            results.append(tensor)
+        await self._transfer(ctx, "READ", transfers)
+        return results
+
+    async def handle_get_request(
+        self,
+        ctx: "TransportContext",
+        entries: list[tuple[Request, Any]],
+    ) -> None:
+        transfers: list[tuple[Request, torch.Tensor, bytes]] = []
```
Comment (ivy-zhou) [inline torchstore/transport/nixl.py:373]: nit: I think you can add a similar comment to handle_put_request here about how this is supposed to be run on the storage volume actor?

File: torchstore/transport/nixl.py [hunk 8]
```diff
@@ -0,0 +1,415 @@
...
+                cache.agent.release_xfer_handle(handle)
+
+    def _connect_client(self, cache: NixlAgentCache) -> str:
+        if self._client_agent_name is None or self._client_metadata is None:
+            raise RuntimeError("NIXL request is missing client agent metadata")
+        return cache.add_remote_agent(
+            self._client_agent_name, self._client_metadata
+        )
+
+    async def handle_put_request(
```
Comment (ivy-zhou) [inline torchstore/transport/nixl.py]: nit: prefer to organize pre-put with put, pre-get with get

File: torchstore/transport/nixl.py [hunk 9]
```diff
@@ -0,0 +1,415 @@
...
+        self,
+        ctx: "TransportContext",
+        entries: list[tuple[Request, Any]],
+    ) -> list[Any]:
+        """Called by storage volume. Read from client's source RdmaMemory (put)."""
+        results: list[Any] = []
+        transfers: list[tuple[Request, torch.Tensor, bytes]] = []
+
+        for (request, tensor), request_context in zip(
```
Comment (ivy-zhou) [inline torchstore/transport/nixl.py:344]: It's probably easier to read if in this and other methods we use the variable name to say where the context is from? For example, this line is `for (request, tensor), client_context` IIUC

File: torchstore/transport/nixl.py [hunk 10]
```diff
@@ -0,0 +1,415 @@
...
+                            "NIXL cannot batch a mix of CPU and GPU tensors on "
+                            "the client side; CPU-to-GPU transfers are supported"
+                        )
+                    for index in range(tensor_local_descs.descCount()):
+                        local_descs.append(tensor_local_descs[index])
+                    for index in range(tensor_remote_descs.descCount()):
+                        remote_descs.append(tensor_remote_descs[index])
+
+            assert local_descs is not None and remote_descs is not None
+            handle = cache.agent.initialize_xfer(
```
Comment (ivy-zhou) [inline torchstore/transport/nixl.py:300]: Codex says that this is an issue if the tensor is a scalar or if the tensor's first dim is 0, because the NIXL Python wrapper can index the tensor without checking its type. Can you double check this please?
```
  For scalar tensors, TorchStore passes the tensor directly here:

  - register_memory(tensor) 
  - Client-side get_xfer_descs(client_tensor) 
  - Volume-side registration and descriptor creation 

  NIXL’s dispatch then does this:

  - get_xfer_descs: accesses descs[0] at line 966 (/data/users/ivyzhou/fbsource/third-party/nixl/v1.3.2/src/src/api/python/_api.py:954), before recognizing a tensor at line 986.
  - get_reg_descs: accesses descs[0] at line 1049 (/data/users/ivyzhou/fbsource/third-party/nixl/v1.3.2/src/src/api/python/_api.py:1037), before recognizing a tensor at line 1069.

  Consequently:

  descs = torch.tensor(1)  # rank 0
  descs[0]                 # IndexError

  The same happens for shapes with an empty first dimension, such as (0,) or (0, 3).
```

File: torchstore/transport/nixl.py [hunk 11]
```diff
@@ -0,0 +1,415 @@
...
+                    serialized_remote_descs
+                )
+
+                if local_descs is None:
+                    local_descs = tensor_local_descs
+                    remote_descs = tensor_remote_descs
+                else:
+                    if local_descs.getType() != tensor_local_descs.getType():
+                        raise RuntimeError(
+                            "NIXL cannot batch a mix of CPU and GPU tensors on "
```
Comment (ivy-zhou) [inline torchstore/transport/nixl.py:286]: Maybe we could split the batch into CPU and GPU groups and then 2 two transfers? I think CPU/CUDA mixed batches are important.

File: torchstore/transport/nixl.py (hunk 10 above)
Comment (ivy-zhou) [reply torchstore/transport/nixl.py:300]: Codex also flagged a similar issue as 0 element tensors, which don't need to be transferred at all, I think we could just have the metadata for them, but this is more of an edge case...

File: torchstore/transport/nixl.py [hunk 12]
```diff
@@ -0,0 +1,415 @@
...
+            if status != "DONE":
+                raise RuntimeError(f"NIXL transfer failed with status {status!r}")
+        except Exception as error:
+            raise RuntimeError(
+                f"NIXL {operation} failed for keys="
+                f"{[request.key for request, _, _ in transfers]!r}"
+            ) from error
+        finally:
+            if handle is not None:
+                cache.agent.release_xfer_handle(handle)
```
Comment (ivy-zhou) [inline torchstore/transport/nixl.py:326]: What if this fails? We probably need to ensure this succeeds? Scenario from Codex:
```
context cleared
      ↓
tensor/storage released
    ↓
PyTorch allocator may reuse address 0x1234
    ↓
still-active NIXL operation writes to 0x1234
    ↓
unrelated new allocation can be overwritten
```

**Review by ivy-zhou — COMMENTED** [review]
Overall looks good to me! Some readability nits and some edge cases that would be good to take a second look at, mainly the CPU + GPU batch thing

File: torchstore/transport/nixl.py (hunk 1 above)
Comment (tushar00jain) [reply torchstore/transport/nixl.py:29]: i'd avoid refactoring existing structure in this diff, we can look at this separately maybe?

File: torchstore/transport/nixl.py (hunk 2 above)
Comment (tushar00jain) [reply torchstore/transport/nixl.py]: does it make a difference?

File: torchstore/transport/nixl.py (hunk 11 above)
Comment (tushar00jain) [reply torchstore/transport/nixl.py]: Sure, had that initially but Titan doesn't need mixed though no harm in supporting it I guess

File: torchstore/transport/nixl.py (hunk 2 above)
Comment (tushar00jain) [reply torchstore/transport/nixl.py]: how about hoisting all env's at the top? i've followed this convention in other projects

File: torchstore/transport/nixl.py (hunk 9 above)
Comment (tushar00jain) [reply torchstore/transport/nixl.py:403]: but everything would just be `client` here?

File: torchstore/transport/nixl.py (hunk 4 above)
Comment (tushar00jain) [reply torchstore/transport/nixl.py:180]: yes i'm putting tests in a separate pr

File: torchstore/transport/nixl.py (hunk 5 above)
Comment (tushar00jain) [reply torchstore/transport/nixl.py:176]: done

File: torchstore/transport/nixl.py (hunk 12 above)
Comment (tushar00jain) [reply torchstore/transport/nixl.py:384]: yes added handling for this

File: torchstore/transport/nixl.py (hunk 3 above)
Comment (tushar00jain) [reply torchstore/transport/nixl.py:430]: done

File: torchstore/transport/nixl.py (hunk 7 above)
Comment (tushar00jain) [reply torchstore/transport/nixl.py:436]: done

File: torchstore/transport/nixl.py (hunk 6 above)
Comment (tushar00jain) [reply torchstore/transport/nixl.py]: done

File: torchstore/transport/nixl.py (hunk 10 above)
Comment (tushar00jain) [reply torchstore/transport/nixl.py]: fixed

File: torchstore/transport/nixl.py (hunk 4 above)
Comment (kapilsh) [reply torchstore/transport/nixl.py:180]: probably ideal if tests land with the PR?

File: torchstore/transport/nixl.py [hunk 13]
```diff
@@ -0,0 +1,470 @@
...
+                        remote_descs,
+                        remote_agent,
+                        backends=[cache.backend],
+                    )
+                    handles.append(handle)
+                    cache.agent.transfer(handle)
+            except Exception as error:
+                # Earlier handles may already be active. Drain them below before
+                # propagating the dispatch error.
+                dispatch_error = error
```
Comment (kapilsh) [inline torchstore/transport/nixl.py:335]: nit: not a huge fan of giant try/catch - can we try to either isolate it to specific failures or if we need it we should move underlyign logic under try to a separate function

File: torchstore/transport/nixl.py [hunk 14]
```diff
@@ -0,0 +1,470 @@
...
+if TYPE_CHECKING:
+    from torchstore.strategy import StorageVolumeRef
+    from torchstore.transport.buffers import TransportContext
+
+try:
+    from nixl import nixl_agent as _nixl_agent
+    from nixl import nixl_agent_config as _nixl_agent_config
+except ImportError:
+    try:
+        from nixl_cu13 import nixl_agent as _nixl_agent
```
Comment (d4l3k) [inline torchstore/transport/nixl.py:38]: it seems like nixl automatically selects the right backend, and these are just legacy -- any reason to keep them?

File: torchstore/transport/nixl.py [hunk 15]
```diff
@@ -0,0 +1,470 @@
...
+
+
+class _NixlTransferStatus(str, Enum):
+    """Status strings returned by NIXL's Python transfer API."""
+    PROCESSING = "PROC"
+    DONE = "DONE"
+
+
+@cache
+def nixl_available() -> bool:
```
Comment (d4l3k) [inline torchstore/transport/nixl.py:67]: I feel like we should detangle available from default. It seems like we should always make it available if nixl is installed and have the flag only control if it's default or not

File: torchstore/transport/nixl.py [hunk 16]
```diff
@@ -0,0 +1,470 @@
...
+
+from torchstore.transport.buffers import TransportBuffer, TransportCache
+from torchstore.transport.types import Request
+
+if TYPE_CHECKING:
+    from torchstore.strategy import StorageVolumeRef
+    from torchstore.transport.buffers import TransportContext
+
+try:
+    from nixl import nixl_agent as _nixl_agent
```
Comment (d4l3k) [inline torchstore/transport/nixl.py:34]: can we avoid importing nixl until we actually instantiate the transport? importing something that isn't used can add significant import times for no benefit

File: torchstore/transport/nixl.py [hunk 17]
```diff
@@ -0,0 +1,470 @@
...
+    PROCESSING = "PROC"
+    DONE = "DONE"
+
+
+@cache
+def nixl_available() -> bool:
+    """Return whether the optional NIXL transport was explicitly enabled."""
+    return (
+        ENV_TORCHSTORE_NIXL_ENABLED
+        and _nixl_agent is not None
```
Comment (d4l3k) [inline torchstore/transport/nixl.py:71]: lets check import metadata rather than importing it for availablity

File: torchstore/transport/nixl.py [hunk 18]
```diff
@@ -0,0 +1,470 @@
...
+    def _evict(self, key: tuple[int, int]) -> None:
+        registration = self._registrations.pop(key, None)
+        if registration is None:
+            return
+        try:
+            self.agent.deregister_memory(
+                registration.descriptors, backends=[self.backend]
+            )
+        except Exception:
+            logger.warning("Failed to deregister NIXL memory", exc_info=True)
```
Comment (d4l3k) [inline torchstore/transport/nixl.py:127]: is this a known issue? why would deregistration fail?

File: torchstore/transport/nixl.py [hunk 19]
```diff
@@ -0,0 +1,470 @@
...
+            self._evict(key)
+        for remote_name in self._remote_agents:
+            try:
+                self.agent.remove_remote_agent(remote_name)
+            except Exception:
+                logger.warning(
+                    "Failed to remove NIXL remote agent %s",
+                    remote_name,
+                    exc_info=True,
+                )
```
Comment (d4l3k) [inline torchstore/transport/nixl.py:158]: Also wondering why we need to handle this case rather than throwing, is this common? do we need to file a bug upstream?

File: torchstore/transport/nixl.py (hunk 13 above)
Comment (tushar00jain) [reply torchstore/transport/nixl.py:335]: same but sometimes really prefer big blocks of code instead of adding small proxy methods - they make you jump around to really understand what's going on

**Review by kapilsh — APPROVED** [review]
Mostly looks fine - let's address existing comments left and let's add tests before landing

File: torchstore/transport/nixl.py (hunk 4 above)
Comment (tushar00jain) [reply torchstore/transport/nixl.py:180]: There's some conventions say put functionality and tests in separate PR and land them separately

There's no risk to landing this because nixl is disabled by default either way

The tests are on the stack so i don't have a preference landing them together or not

File: torchstore/transport/nixl.py (hunk 9 above)
Comment (ivy-zhou) [reply torchstore/transport/nixl.py:391]: Yes, I just mean for consistency, it's nice to write client_request_context here

**Review by ivy-zhou — APPROVED** [review]
LGTM, ptal at @d4l3k 's comments before landing 👍

File: torchstore/transport/nixl.py (hunk 15 above)
Comment (tushar00jain) [reply torchstore/transport/nixl.py:66]: The current transports all follow the same convention right now, maybe we can consider doing this refactor later?

File: torchstore/transport/nixl.py (hunk 9 above)
Comment (tushar00jain) [reply torchstore/transport/nixl.py:391]: sorry you mean consistent with what? request_context just represents `NixlRequestContext` - there's no client in the name
